skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+        intervals : an iterable of ``Interval`` objects
+        """
+        if metadata is None:
+            return
+            yield
+
+        if intervals is None:
+            intervals = self._intervals
+
+        for intvl in intervals:
+            for (key, value) in metadata.items():
+                if (key not in intvl.metadata or
+                        intvl.metadata[key] != value):
+                    break
+            else:
+                yield intvl
+
+    @experimental(as_of='0.5.0-dev')
+    @_rebuild_tree
+    def query(self, locations=None, metadata=None):
+        """Yield ``Interval`` object with the locations and attributes.
+
+        The ``Interval`` objects must meet both requirements: 1) overlap
+        with any of the spans specified by ``locations``; 2) satisfy
+        ``metadata`` specification. For instance, you can identify
+        all the recA genes that overlap with (10, 100) or (900, 1000)
+        with this code ``interval_metadata.query([(10, 100),
+        (900, 1000)], {'gene': 'recA'})``.
+
+        Parameters
+        ----------
+        locations : iterable of tuples of int pair
+            Specifies locations to look for the ``Interval``
+            objects. An satisfying interval feature only need to overlap with
+            one location. Default (``None``) means all ``Interval``s meet
+            this requirement.
+
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            ``Interval`` object. It specifies what metadata keywords and
+            values to look for. Default (``None``) means all ``Interval``s
+            meet this requirement.
+
+        Returns
+        -------
+        generator, Interval
+            A generator of Interval objects satisfying the search criteria.
+        """
+        if locations is None:
+            for intvl in self._query_attribute(metadata):
+                yield intvl
+        else:
+            for loc in locations:
+                intvls = self._query_interval(loc)
+                if metadata is None:
+                    metadata = {}
+                for intvl in self._query_attribute(metadata, intvls):
+                    yield intvl
+
+    @experimental(as_of='0.5.0-dev')
+    def drop(self, locations=None, metadata=None):
+        """Drops Interval objects according to a specified query.
+
+        Locations are first queried from the ``IntervalMetadata`` object
+        using the query functionality. These locations are then dropped.
+
+        Parameters
+        ----------
+        locations : iterable of tuple of ints
+            A list of locations associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+        """
+        to_delete = {id(f) for f in
+                     self.query(locations=locations,
+                                metadata=metadata)}
```

Comment:
members of `to_delete` never learn of their dropped status. We'll need to let them know that their `interval_metadata` is `None`